### DISCUSSION TITLE
Any way to have more than 1 glyph map for the same input device, and choose which one to use?

### DISCUSSION THREAD

**Discussion by TheEmbracedOne** [Orange-Panda/Rewired-Glyphs#12] [Q&A] [answered]
Hi, it's me again, with yet another question or possibly feature request.
I would like to make multiple glyph maps for the same input device and choose within the <Glyph> tag which one to use.

My use case is that there are different contexts where I would like to sometimes display glyphs with a background (the back tilted squares in this case):
![image](https://github.com/Orange-Panda/Rewired-Glyphs/assets/16197906/81aa0479-4939-488f-9e5d-609d82ad969c)

while in some places, without a background:
![Unity_1ZoeUcVDV4](https://github.com/Orange-Panda/Rewired-Glyphs/assets/16197906/54cf33e1-a004-4e3e-90d2-a373a3d199da)

Up until now, I faked these backgrounds by putting an image behind my glyohs, but most of my use cases, I use Glyphs with text, so it's hard to align them. This is why I thought that perhaps having multiple glpyh maps could be the solution, so I could make a glyph map with the same icons where the background would be "baked in" into the glyph texture..

So in this example, I'd like to have two playstation glyph maps, and depending on context, I'd either use `<Glyph Submit type=bg>` for the one with backgrounds, and `<Glyph Submit type=nobg>` for the one without backgrounds:

![Unity_RMPAub20nu](https://github.com/Orange-Panda/Rewired-Glyphs/assets/16197906/2b083f61-01f5-4f78-b8eb-dc4b87d97441)

This could also benefit a use case where a game had a setting for light or dark glyphs (not my use case, but thought I'd mention another use).

**Comment by Orange-Panda** [comment]
There is support for loading an entirely separate set of glyphs using `InputGlyphs.LoadGlyphCollection` which would be perfect for the light mode/dark mode set of glyphs use case you describe.

However, I don't think having multiple styles in a single glyph collection will be a feasible addition (at least with the current infrastructure). This is something I will consider but I don't believe this can be added without really significant changes unfortunately

**Reply by Orange-Panda** [reply]
Just to clarify I will be investigating this addition further but I can't promise anything surrounding it

**Reply by TheEmbracedOne** [reply]
Oh, if I could use `InputGlyphs.LoadGlyphCollection` to load a different glyph map, that'd be good for me and my use case.

I created another GlyphCollection by duplicating the one I was using, duplicated the ps5 glyph map, replaced one glyph with an image with a baked in background just for testing, then put it on a text object with the following script:
```csharp
	public class GlyphTest : MonoBehaviour {
		public GlyphCollection bgCollection;
		public GlyphCollection noBgCollection;

		[Button("LoadBackgroundGlyphs")]
		public void LoadBackgroundGlyphs() { InputGlyphs.LoadGlyphCollection(bgCollection); }

		[Button("LoadNoBackgroundGlyphs")]
		public void LoadNoBackgroundGlyphs() { InputGlyphs.LoadGlyphCollection(noBgCollection); }
	}
```
however, when loading the background glyphs, nothing changes (using NaughtyAttributes, I can press a button in the inspector to trigger those functions). I know I made a mistake here, and I immediately thought that maybe I need to do something with the RewiredGlyphManager to load the right GlyphCollection, since it still has the old one (well, nothing "told" it to load another one), and there is no TMP Sprite Sheet generated of the background version of the Glyph Collection, it'd seem to me that if I swapped out the GlyphCollection on the RewiredGlyphManager and pressed the "Generate TMP Sprite Sheet", it'd generate a new sprite sheet, overwriting the old one?

I'd appreciate some guidance so I can figure this out! Thank you

**Reply by TheEmbracedOne** [reply]
Followup to my last post, I have been able to get it to work in the end I think, or at least, had a breakthrough:
![Unity_9HF3gHIjYi](https://github.com/Orange-Panda/Rewired-Glyphs/assets/16197906/ea05c67f-627b-4ccf-9a6b-3644f42d4b6c)

The GlyphTest does the following:
```csharp
	public class GlyphTest : MonoBehaviour {
		public RewiredGlyphManager glyphManager;
		public GlyphCollection bgCollection;
		public GlyphCollection noBgCollection;

		[Button("LoadBackgroundGlyphs")]
		public void LoadBackgroundGlyphs() { glyphManager.UpdateGlyphCollection(bgCollection); }

		[Button("LoadNoBackgroundGlyphs")]
		public void LoadNoBackgroundGlyphs() { glyphManager.UpdateGlyphCollection(noBgCollection); }
	}
```

and I added within the RewiredGlyphManager the following bit:
```csharp
		public void UpdateGlyphCollection(GlyphCollection _newGlyphCollection) {
			glyphCollection = _newGlyphCollection;
			InputGlyphs.LoadGlyphCollection(glyphCollection);
		}
```

After creating both Glyph Collections, I made sure that the assets they use do not have the same names, adding a "_bg" prefix ensured that they do not overwrite each other. Then, I plugged each Glyph Collection into RewiredGlyphManager to generate their TMP Sprite Sheets. Seems it was enough to make it work, I'll do some more testing.

**Reply by Orange-Panda** [reply]
- Is this load function happening at runtime?
- The field on RewiredGlyphManager occurs once when it starts and loads that collection (essentially as the "default" collection). From that point onward you can load another collection to replace it.
- Regarding sprite sheets: That is correct that you would have to swap them out on the manager temporarily and generate sprite sheets for every collection. However, they shouldn't overwrite each other unless they have identical names. If they are overwriting each other that may have been a design oversight on my part that I'll have to look into
- Something you should beware of is that the collection is global so it will affect **all** glyph outputs.
- The load method is also moderately expensive so avoid calling it every frame. Your example doesn't but just emphasizing that fact just in case.

**Reply by TheEmbracedOne** [reply]
* Is this load function happening at runtime? - Yes, the idea would be that when I enter a screen where I need to show Glyphs without a background, I'd switch to the other GlyphCollection, and when I exit that screen, switch back to the collection with backgrounds, as most of my use case is showing them with backgrounds.
* Regarding sprite sheets: I realised this after some testing, and I just added _bg suffix to glyphs with backgrounds, and now it works!

That said, the more I think about it and the more I was testing this, I wonder if I should just simplify things and decide to run with a Glyph Collection that uses glyphs always with a background. Perhaps switching between these is not as great of an idea as I originally thought. I'll do some further experiments to see if it's worth the trouble!

**Reply by Orange-Panda** [reply]
Additionally I reflect on the design of the collection loading and realized the program could potentially have multiple collections in memory at once (having one as the default matching current behavior) and add functionality that tries to get glyphs from a different non-default collection in memory. That would essentially function as the original request where you could switch styles on a per output basis and would seemingly be backwards compatible too. 

I'll have to experiment with this idea so still no promises but it is more feasible than I originally thought.

**Reply by Orange-Panda** [reply]
Update: I was able to [implement the process of loading several collections at once](https://github.com/Orange-Panda/Rewired-Glyphs/commit/fa7d90c8acdd2556d0d2149487c36c4f01a12d8c) and sorting them by a string key so I feel comfortable saying this feature is planned to be fully implemented. 

The internal changes that made this possible have also caused loading collections to be significantly more performant since it only initializes it once (before it was initializing it every time it is loaded).

There is still work to be done making the access of the secondary collections possible publicly but wanted to share this update to keep you in the loop on progress.

**Reply by TheEmbracedOne** [reply]
Thank you for the update, and good work, wow!

**Reply by Orange-Panda** [reply, marked as answer]
This feature is now available with the [2.0.0](https://github.com/Orange-Panda/Rewired-Glyphs/discussions/13) release. 

A summary of the implementation:
- GlyphCollection now has a `key` value that is used for identifying it
- You can now load multiple collections on start using the `Additional Collections` field on the `RewiredGlyphManager`
- Most glyph queries now have a `collectionKey` parameter which is used for targeting secondary collections
- You can also target secondary collections in glyph rich text using the `set=` parameter such as `set=dark` where `dark` is the key of some GlyphCollection
- The other built-in components *do not* yet support targeting secondary collections as of `2.0.0` but this may be implemented if desired
- The [LoadGlyphCollection](https://orange-panda.github.io/Rewired-Glyphs/api/LMirman.RewiredGlyphs.InputGlyphs.LoadGlyphCollection.html) method is still available and is now used for loading a collection and setting it as either the primary or secondary collection

**Reply by TheEmbracedOne** [reply]
Godly. Thank you so much! I'm grabbing the update and will get to work right away!
